File: app/src/main/java/com/galtashma/parsedashboard/adapters/ParseObjectsAdapter.java
package com.galtashma.parsedashboard.adapters;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;

import com.galtashma.lazyparse.LazyList;
import com.galtashma.lazyparse.LazyParseObjectHolder;
import com.galtashma.lazyparse.ScrollInfiniteAdapter;
import com.galtashma.parsedashboard.R;
import com.lucasurbas.listitemview.ListItemView;
import com.parse.ParseObject;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by gal on 3/14/18.
 */

public class ParseObjectsAdapter extends ScrollInfiniteAdapter<ParseObject> {

    private List<String> previewFieldNames;

    public ParseObjectsAdapter(Context context, LazyList<ParseObject> lazyValues) {
        super(context, lazyValues, R.layout.list_item, 15);
        previewFieldNames = Arrays.asList("createdAt", "updatedAt");
    }

    public ParseObjectsAdapter(Context context, LazyList<ParseObject> lazyValues, List<String> previewFields) {
        super(context, lazyValues, R.layout.list_item, 15);
        this.previewFieldNames = previewFields;
    }

    public void updatePreviewFields(List<String> previewFields){
        this.previewFieldNames = previewFields;
        this.notifyDataSetChanged();
    }

    @Override
    public View renderReadyLazyObject(ParseObject t, View view, @NonNull ViewGroup viewGroup) {
        ListItemView item = (ListItemView) view;
        item.setTitle(t.getObjectId());
        item.setMultiline(true);

        Map<String, String> fields = new HashMap<>();
        for (String fieldName : previewFieldNames) {
            String value = formatParseField(t, fieldName);
            if (value != null) {
                fields.put(fieldName, formatParseField(t, fieldName));
            }
        }
        item.setSubtitle(mapToString(fields));
        return view;
    }

    private String formatParseField(ParseObject t, String key) {
        if (key.equals("createdAt")) {
            return t.getCreatedAt().toString();
        }
        if (key.equals("updatedAt")) {
            return t.getCreatedAt().toString();
        }

        if (t.has(key)) {
            return t.get(key).toString();
        }

        return null;
    }

    @Override
    public View renderLoadingLazyObject(LazyParseObjectHolder<ParseObject> lazyParseObjectHolder, View view, @NonNull ViewGroup viewGroup) {
        ListItemView item = (ListItemView) view;
        item.setTitle("Loading...");
        return view;
    }

    private String mapToString(Map<String, String> map) {
        if (map.size() <= 0) {
            return "";
        }
        
        StringBuilder sb = new StringBuilder();
        for (String k : map.keySet()) {
            sb.append(k);
            sb.append(": ");
            sb.append(map.get(k));
            sb.append("\n");
        }

        sb.delete(sb.length()-1, sb.length()); // remove last \n

        return sb.toString();
    }
}

File: app/src/main/java/com/galtashma/parsedashboard/adapters/ParseObjectsAdapter.kt
package com.galtashma.parsedashboard.adapters

import android.content.Context
import android.view.View
import android.view.ViewGroup
import androidx.annotation.NonNull
import com.galtashma.lazyparse.LazyList
import com.galtashma.lazyparse.LazyParseObjectHolder
import com.galtashma.lazyparse.ScrollInfiniteAdapter
import com.galtashma.parsedashboard.R
import com.lucasurbas.listitemview.ListItemView
import com.parse.ParseObject

/**
 * Created by gal on 3/16/18, rewritten by Cyb3rKo on 05/12/22.
 */

class ParseObjectsAdapter(
    context: Context,
    lazyValues: LazyList<ParseObject>,
    private var previewFieldNames: List<String>
) : ScrollInfiniteAdapter<ParseObject>(context, lazyValues, R.layout.list_item, 15) {

    fun updatePreviewFields(previewFields: List<String>){
        previewFieldNames = previewFields
        this.notifyDataSetChanged()
    }

    override fun renderReadyLazyObject(
        parseObject: ParseObject,
        view: View?,
        @NonNull viewGroup: ViewGroup
    ): View {
        val item = view as ListItemView
        item.title = parseObject.objectId
        item.setMultiline(true)

        val fields = HashMap<String, String?>()
        for (fieldName in previewFieldNames) {
            val value = formatParseField(parseObject, fieldName)
            if (value != null) {
                fields[fieldName] = formatParseField(parseObject, fieldName)
            }
        }
        item.subtitle = mapToString(fields)
        return view
    }

    private fun formatParseField(parseObject: ParseObject, key: String): String? {
        if (key == "createdAt") {
            return parseObject.createdAt.toString()
        }
        if (key == "updatedAt") {
            return parseObject.createdAt.toString()
        }

        if (parseObject.has(key)) {
            return parseObject.get(key).toString()
        }

        return null
    }

    override fun renderLoadingLazyObject(
        lazyParseObjectHolder: LazyParseObjectHolder<ParseObject>?,
        view: View?,
        viewGroup: ViewGroup
    ): View {
        val item = view as ListItemView
        item.title = "Loading..."
        return view
    }

    private fun mapToString(map: Map<String, String?>): String {
        if (map.isEmpty()) return ""

        val sb = StringBuilder()
        for (k in map.keys) {
            sb.append("$k: ${map[k]}\n")
        }

        sb.dropLast(1) // remove last \n

        return sb.toString()
    }
}
